fix: Update tx flow form when selecting a replacement nonce

usame-algan opened the pull request:
## What it solves

Part of #2383 

The form mode for the nonce form changed to `onTouched` with #2383 but selecting a nonce now doesn't update the form unless the user blurs the nonce input. This change calls `onBlur()` whenever the user selects something from the nonce replacement dropdown.

## How to test it

1. Open a safe with at least one transaction in queue
2. Create a new transaction
3. Go to the review screen
4. Observe seeing the Sign flow
5. Select the existing nonce from the nonce form
6. Observe seeing the Execute flow

## Checklist
* [ ] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [ ] I've written a unit/e2e test for it (if applicable) 🧑‍💻

katspaugh reviewed (approved):
Good catch!

schmanu reviewed (approved):
💯 Nice one!

francovenica commented:
So the main flow works fine, if the nonce is the next one to be executed the form allows execution, if is not only signing is possible.

Small issue:
There used to be a skeleton in the input that allow the load of the nonce without flickering. In this PR that skeleton is missing, so you can see the lowest nonces for a couple of seconds until it changes to the next nonce in the queue:
![nonce input](https://github.com/safe-global/safe-wallet-web/assets/4503659/ccfa08a1-9766-4624-8e0f-b42e9bcf98c0)

usame-algan commented:
I've reverted part of #2383 to have the previous validation UX (checked without having to blur the input) and instead added an effect to update the form value when the recommended nonce changes and IF the user has not touched the field yet. I've also added back the loading state. @schmanu @katspaugh can you check if this is the desired behaviour?

katspaugh commented on src/components/tx-flow/SafeTxProvider.tsx:
```diff
@@ -42,11 +42,12 @@ const SafeTxProvider = ({ children }: { children: ReactNode }): ReactElement =>
   // Recommended nonce and safeTxGas
-  const recommendedNonce = Math.max(safe.nonce, useRecommendedNonce() ?? 0)
+  const recommendedNonce = useRecommendedNonce()
+  const safeRecommendedNonce = recommendedNonce ? Math.max(safe.nonce, recommendedNonce) : undefined
```
IMHO this should be part of useRecommendedNonce

usame-algan replied:
True, I've moved it into the hook.

katspaugh commented on src/components/tx-flow/common/TxNonce/index.tsx:
```diff
@@ -104,12 +104,20 @@ const TxNonceForm = ({ nonce, recommendedNonce }: { nonce: string; recommendedNo
+  useEffect(() => {
+    if (formMethods.formState.touchedFields[TxNonceFormFieldNames.NONCE]) return
+
+    formMethods.setValue(TxNonceFormFieldNames.NONCE, recommendedNonce)
```
Why is needed? Can’t the input just be a controlled input?

usame-algan replied:
Good point, I think the issue is that the `validate` calls `setNonce` automatically without user input so that the controlled input value `nonce` will never update to the recommended nonce. I've added an early return inside the `validate` function to skip validation if the input value is the same as `nonce` to avoid initial validation. This fixes the issue too and since we know that `nonce` is always valid it is ok to skip validation if the input value is the same.

schmanu reviewed (approved):
I tried it out and it works fine.

francovenica commented:
Looks good to me.
The skeleton to hide the nonce loading is working fine
Changing the nonce (by writing, by selecting from the dropdown or using the circular arrow) changes the form to execution or only-siging just fine.

katspaugh commented on src/components/tx/SignOrExecuteForm/hooks.ts:
```diff
@@ -165,10 +165,12 @@ export const useRecommendedNonce = (): number | undefined => {
-      return getRecommendedNonce(safe.chainId, safeAddress)
+      const recommendedNonce = await getRecommendedNonce(safe.chainId, safeAddress)
+
+      return recommendedNonce ? Math.max(safe.nonce, recommendedNonce) : undefined
```
Please move this to `useRecommendedNonce`.
